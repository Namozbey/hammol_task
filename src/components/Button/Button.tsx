/* eslint-disable indent */
import React, { ReactNode } from "react";
import styled from "styled-components";

interface Props {
  type?: "button" | "submit" | "reset";
  children: number | string | ReactNode;
  variant: "text" | "contained" | "outlined";
  size: {
    x: string;
    y: string;
  };
  icon?: {
    left?: string;
    right?: string;
  };
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

const defineColorByVariant = (variant: string) => {
  switch (variant) {
    case "text":
      return `
        background-color: rgba(96, 96, 96, 0.68);
        color: #f65261;
        border-color: transparent;
      `;
    case "contained":
      return `
        background-color: #f65261;
        color: #fff;
        border-color: #f65261;
      `;
    case "outlined":
      return `
        color: #f65261;
        background-color: transparent;
        border-color: #f65261;
      `;
    default:
      return "";
  }
};

const StyledButton = styled.button<Props>`
  width: ${(props) => props.size.x};
  height: ${(props) => props.size.y};
  ${(props) => defineColorByVariant(props.variant)}
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 4px;
  line-height: 24px;
  font-size: 20px;
  text-transform: uppercase;
  font-family: Montserrat, sans-serif;
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  -webkit-touch-callout: none;
  border-style: solid;
  border-width: 1.5px;
  margin: 0;
  padding: 0;

  &:active {
    opacity: 0.8;
    transition-duration: 150ms;
  }
`;
export default function Button(props: Props): JSX.Element {
  const { type, children, size, variant, icon, ...rest } = props;

  return (
    <StyledButton
      size={size}
      type={type}
      variant={variant}
      // className={`nr-button nr-button-${variant}`}
      {...rest}
    >
      {icon && icon.left && <span className="material-icons">{icon.left}</span>}
      {children}
      {icon && icon.right && (
        <span className="material-icons">{icon.right}</span>
      )}
    </StyledButton>
  );
}

Button.defaultProps = {
  children: "",
  variant: "contained",
  size: { x: "180px", y: "57px" },
  type: "button",
};
