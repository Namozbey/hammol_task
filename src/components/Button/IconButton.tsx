import React, { ReactNode } from "react";
import styled from "styled-components";

interface Props {
  style?: object;
  className?: string;
  iconClassName?: string;
  iconName?: string;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  [key: string]: string | number | ReactNode;
}

const StyledIconButton = styled.button`
  width: 36px;
  height: 36px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  margin: 0;
  padding: 0;
  border: none;
  cursor: pointer;
  background-color: rgba(35, 35, 35, 0.9);

  > span {
    color: #fff;
  }

  &:active {
    opacity: 0.8;
    transition-duration: 150ms;
  }
`;

export default function IconButton(props: Props): JSX.Element {
  const { style, className, iconClassName, iconName, onClick, ...rest } = props;
  return (
    <StyledIconButton
      style={style}
      className={className}
      onClick={onClick}
      {...rest}
    >
      <span className={iconClassName}>{iconName}</span>
    </StyledIconButton>
  );
}

IconButton.defaultProps = {
  iconClassName: "material-icons",
  iconName: "more_vert",
};
