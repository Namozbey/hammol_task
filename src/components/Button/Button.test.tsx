import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Button from "./Button";

describe("Testing Button", () => {
  it("Snapshots", () => {
    const ContainedButton = render(
      <Button variant="contained">contained</Button>
    );
    expect(ContainedButton.asFragment()).toMatchSnapshot();

    const OutlinedButton = render(<Button variant="outlined">outlined</Button>);
    expect(OutlinedButton.asFragment()).toMatchSnapshot();

    const textButton = render(<Button variant="text">text</Button>);
    expect(textButton.asFragment()).toMatchSnapshot();
  });

  it("Fire click event", () => {
    const onClick = (e: React.MouseEvent<HTMLButtonElement>) => {
      expect(e).not.toBeUndefined();
      expect(e).not.toBeNull();
    };

    const { getByText } = render(<Button onClick={onClick}>click</Button>);
    const element = getByText("click");
    fireEvent.click(element);
  });
});
