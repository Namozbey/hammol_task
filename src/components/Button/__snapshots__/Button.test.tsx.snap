// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Button Snapshots 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ ePqhfu"
    type="button"
  >
    contained
  </button>
</DocumentFragment>
`;

exports[`Testing Button Snapshots 2`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ iwbzml"
    type="button"
  >
    outlined
  </button>
</DocumentFragment>
`;

exports[`Testing Button Snapshots 3`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ cQlmAu"
    type="button"
  >
    text
  </button>
</DocumentFragment>
`;
